### DISCUSSION TITLE
Error reading Landsat 9 files

### DISCUSSION THREAD

**Discussion by cullen-molitor** [microsoft/PlanetaryComputer#101] [Q&A]
I am running through a loop which pulls the least cloudy Landsat image from the collection in a given time range. Some of these time ranges return Landsat 9 imagery. I am running code in [this notebook](https://github.com/cropmosaiks/crop-modeling/blob/main/code/2_feature_extraction/s-2_ls-c2-l2_multiband.ipynb). Here is the error I get for the least cloudy image of July 2022 which covers `array([ 24.784878, -16.664232])`:

<details>
<summary>
Error
</summary>

```python
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
File <timed exec>:189, in <module>

File /srv/conda/envs/notebook/lib/python3.8/site-packages/torch/utils/data/dataloader.py:521, in _BaseDataLoaderIter.__next__(self)
    519 if self._sampler_iter is None:
    520     self._reset()
--> 521 data = self._next_data()
    522 self._num_yielded += 1
    523 if self._dataset_kind == _DatasetKind.Iterable and \
    524         self._IterableDataset_len_called is not None and \
    525         self._num_yielded > self._IterableDataset_len_called:

File /srv/conda/envs/notebook/lib/python3.8/site-packages/torch/utils/data/dataloader.py:1183, in _MultiProcessingDataLoaderIter._next_data(self)
   1181 if len(self._task_info[self._rcvd_idx]) == 2:
   1182     data = self._task_info.pop(self._rcvd_idx)[1]
-> 1183     return self._process_data(data)
   1185 assert not self._shutdown and self._tasks_outstanding > 0
   1186 idx, data = self._get_data()

File /srv/conda/envs/notebook/lib/python3.8/site-packages/torch/utils/data/dataloader.py:1229, in _MultiProcessingDataLoaderIter._process_data(self, data)
   1227 self._try_put_index()
   1228 if isinstance(data, ExceptionWrapper):
-> 1229     data.reraise()
   1230 return data

File /srv/conda/envs/notebook/lib/python3.8/site-packages/torch/_utils.py:434, in ExceptionWrapper.reraise(self)
    430 except TypeError:
    431     # If the exception takes multiple arguments, don't try to
    432     # instantiate since we don't know how to
    433     raise RuntimeError(msg) from None
--> 434 raise exception

RuntimeError: Caught RuntimeError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "rasterio/_base.pyx", line 261, in rasterio._base.DatasetBase.__init__
  File "rasterio/_shim.pyx", line 78, in rasterio._shim.open_dataset
  File "rasterio/_err.pyx", line 216, in rasterio._err.exc_wrap_pointer
rasterio._err.CPLE_OpenFailedError: '/vsicurl/https://landsateuwest.blob.core.windows.net/landsat-c2/level-2/standard/oli-tirs/2022/174/072/LC09_L2SP_174072_20220705_20220803_02_T1/LC09_L2SP_174072_20220705_20220803_02_T1_SR_B7.TIF?st=2022-09-05T20%3A58%3A32Z&se=2022-09-13T20%3A58%3A32Z&sp=rl&sv=2021-06-08&sr=c&skoid=c85c15d6-d1ae-42d4-af60-e2ca0f81359b&sktid=72f988bf-86f1-41af-91ab-2d7cd011db47&skt=2022-09-06T20%3A58%3A31Z&ske=2022-09-13T20%3A58%3A31Z&sks=b&skv=2021-06-08&sig=KCftSi0K6bUKiPE9cPkEUcaXLng9b5T5nEXaYN5GySo%3D' not recognized as a supported file format.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/stackstac/rio_reader.py", line 326, in _open
    ds = SelfCleaningDatasetReader(
  File "rasterio/_base.pyx", line 263, in rasterio._base.DatasetBase.__init__
rasterio.errors.RasterioIOError: '/vsicurl/https://landsateuwest.blob.core.windows.net/landsat-c2/level-2/standard/oli-tirs/2022/174/072/LC09_L2SP_174072_20220705_20220803_02_T1/LC09_L2SP_174072_20220705_20220803_02_T1_SR_B7.TIF?st=2022-09-05T20%3A58%3A32Z&se=2022-09-13T20%3A58%3A32Z&sp=rl&sv=2021-06-08&sr=c&skoid=c85c15d6-d1ae-42d4-af60-e2ca0f81359b&sktid=72f988bf-86f1-41af-91ab-2d7cd011db47&skt=2022-09-06T20%3A58%3A31Z&ske=2022-09-13T20%3A58%3A31Z&sks=b&skv=2021-06-08&sig=KCftSi0K6bUKiPE9cPkEUcaXLng9b5T5nEXaYN5GySo%3D' not recognized as a supported file format.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/torch/utils/data/_utils/worker.py", line 287, in _worker_loop
    data = fetcher.fetch(index)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py", line 49, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py", line 49, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "<timed exec>", line 149, in __getitem__
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/xarray/core/dataarray.py", line 947, in compute
    return new.load(**kwargs)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/xarray/core/dataarray.py", line 921, in load
    ds = self._to_temp_dataset().load(**kwargs)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/xarray/core/dataset.py", line 861, in load
    evaluated_data = da.compute(*lazy_data.values(), **kwargs)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/dask/base.py", line 575, in compute
    results = schedule(dsk, keys, **kwargs)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/dask/local.py", line 554, in get_sync
    return get_async(
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/dask/local.py", line 497, in get_async
    for key, res_info, failed in queue_get(queue).result():
  File "/srv/conda/envs/notebook/lib/python3.8/concurrent/futures/_base.py", line 437, in result
    return self.__get_result()
  File "/srv/conda/envs/notebook/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/dask/local.py", line 539, in submit
    fut.set_result(fn(*args, **kwargs))
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/dask/local.py", line 235, in batch_execute_tasks
    return [execute_task(*a) for a in it]
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/dask/local.py", line 235, in <listcomp>
    return [execute_task(*a) for a in it]
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/dask/local.py", line 226, in execute_task
    result = pack_exception(e, dumps)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/dask/local.py", line 221, in execute_task
    result = _execute_task(task, data)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/dask/core.py", line 119, in _execute_task
    return func(*(_execute_task(a, cache) for a in args))
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/dask/core.py", line 119, in <genexpr>
    return func(*(_execute_task(a, cache) for a in args))
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/dask/core.py", line 119, in _execute_task
    return func(*(_execute_task(a, cache) for a in args))
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/dask/optimization.py", line 990, in __call__
    return core.get(self.dsk, self.outkey, dict(zip(self.inkeys, args)))
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/dask/core.py", line 149, in get
    result = _execute_task(task, cache)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/dask/core.py", line 119, in _execute_task
    return func(*(_execute_task(a, cache) for a in args))
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/stackstac/to_dask.py", line 185, in fetch_raster_window
    data = reader.read(current_window)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/stackstac/rio_reader.py", line 385, in read
    reader = self.dataset
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/stackstac/rio_reader.py", line 381, in dataset
    self._dataset = self._open()
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/stackstac/rio_reader.py", line 337, in _open
    raise RuntimeError(msg) from e
RuntimeError: Error opening 'https://landsateuwest.blob.core.windows.net/landsat-c2/level-2/standard/oli-tirs/2022/174/072/LC09_L2SP_174072_20220705_20220803_02_T1/LC09_L2SP_174072_20220705_20220803_02_T1_SR_B7.TIF?st=2022-09-05T20%3A58%3A32Z&se=2022-09-13T20%3A58%3A32Z&sp=rl&sv=2021-06-08&sr=c&skoid=c85c15d6-d1ae-42d4-af60-e2ca0f81359b&sktid=72f988bf-86f1-41af-91ab-2d7cd011db47&skt=2022-09-06T20%3A58%3A31Z&ske=2022-09-13T20%3A58%3A31Z&sks=b&skv=2021-06-08&sig=KCftSi0K6bUKiPE9cPkEUcaXLng9b5T5nEXaYN5GySo%3D': RasterioIOError("'/vsicurl/https://landsateuwest.blob.core.windows.net/landsat-c2/level-2/standard/oli-tirs/2022/174/072/LC09_L2SP_174072_20220705_20220803_02_T1/LC09_L2SP_174072_20220705_20220803_02_T1_SR_B7.TIF?st=2022-09-05T20%3A58%3A32Z&se=2022-09-13T20%3A58%3A32Z&sp=rl&sv=2021-06-08&sr=c&skoid=c85c15d6-d1ae-42d4-af60-e2ca0f81359b&sktid=72f988bf-86f1-41af-91ab-2d7cd011db47&skt=2022-09-06T20%3A58%3A31Z&ske=2022-09-13T20%3A58%3A31Z&sks=b&skv=2021-06-08&sig=KCftSi0K6bUKiPE9cPkEUcaXLng9b5T5nEXaYN5GySo%3D' not recognized as a supported file format.")
```

</details>

Any idea what is causing this? This is the second time it has come up and both times it was for Landsat 9 imagery

**Comment by TomAugspurger** [comment]
Thanks for reporting this. It seems that the USGS server returned a bad result, and our data pipeline didn't handle it properly. The actual contents of that file are HTML

Simplifying things a bit:


```python
import planetary_computer
import requests
from IPython.display import HTML
url = "https://landsateuwest.blob.core.windows.net/landsat-c2/level-2/standard/oli-tirs/2022/174/072/LC09_L2SP_174072_20220705_20220803_02_T1/LC09_L2SP_174072_20220705_20220803_02_T1_SR_B7.TIF"
url = planetary_computer.sign(url)
HTML(requests.get(url).text)
```

<img width="1806" alt="image" src="https://user-images.githubusercontent.com/1312546/188749229-057a22dc-c467-447d-882e-3a4706f2673f.png">

We'll look into it (along with searching for other places this might have happened)

**Reply by cullen-molitor** [reply]
Found another  one at `https://landsateuwest.blob.core.windows.net/landsat-c2/level-2/standard/oli-tirs/2022/172/067/LC09_L2SP_172067_20220402_20220404_02_T1/LC09_L2SP_172067_20220402_20220404_02_T1_SR_B6.TIF`

**Reply by cullen-molitor** [reply]
It looks like the problem may apply to Landsat 8 imagery in the new collection as well. I ran into the problem with this file `https://landsateuwest.blob.core.windows.net/landsat-c2/level-2/standard/oli-tirs/2022/172/072/LC08_L2SP_172072_20220512_20220519_02_T1/LC08_L2SP_172072_20220512_20220519_02_T1_SR_B5.TIF`

**Comment by cullen-molitor** [comment]
Any movement on reprocessing these files?

**Reply by TomAugspurger** [reply]
Some. The problem we've run into is that USGS has apparently removed some of original scenes, some of which have been replaced by new scenes (reprocessed) but others haven't. We're trying to track those down before figuring out alternatives.

**Reply by TomAugspurger** [reply]
Quick status update here: we've got a list of scenes with the bad TIFF files. USGS has reprocessed many of those scenes (and deleted the original data). I'm able to get a list of the matching scenes and will begin to copy the data over once USGS's distribution service has stabilized a bit. It will take some time for all of the bytes to be copied.

The STAC item IDs will not change (we don't include the processing timestamp in our item IDs). The assets for that scene will be all new, and the original assets will be deleted.

**Comment by jessjaco** [comment]
It's probably on your list but here is another file with this issue: https://landsateuwest.blob.core.windows.net/landsat-c2/level-2/standard/oli-tirs/2022/101/055/LC08_L2SP_101055_20220612_20220617_02_T2/LC08_L2SP_101055_20220612_20220617_02_T2_SR_B3.TIF

Trying to get stackstac or ods.stac to ignore files like this has proven difficult (using e.g. 'errors_as_nodata' for stacstack.stack doesn't seem to working in this case). Short of trying to open every asset and catching errors before calling stack, I've resorted to maintaining a list of bad item ids to filter the itemcollection by prior to reading. Anyone have alternative ideas?

**Reply by TomAugspurger** [reply]
That scene is indeed on the list. We're currently paused since USGS is offline for maintenance. It'll pick back up again tonight or tomorrow.

That said, we are seeing a number of "fixed" scenes missing from USGS's download endpoint. I'm waiting to hear back about those.

**Reply by jessjaco** [reply]
We've resorted to establishing our own list of bad ids and filtering them out of any item collections. Would it be possible to get access the full list that you created?

**Reply by TomAugspurger** [reply]
I suspect that list won't be too useful now, since it's now a mixture of items that were fixed with this reingestion (which just completed) and items that failed to reingest.

I'll post an update once I hear back from USGS, and we'll work towards getting every single item fixed.

**Comment by mikemahoney218** [comment]
Just in case it's not on the list, wanted to flag https://landsateuwest.blob.core.windows.net/landsat-c2/level-2/standard/oli-tirs/2022/014/031/LC08_L2SP_014031_20220626_20220706_02_T1/LC08_L2SP_014031_20220626_20220706_02_T1_SR_B3.TIF as the same issue :smile: 

Looking forward to an update when this is fixed!

**Reply by TomAugspurger** [reply]
That is on the list, and was unfortunately one of the ones that failed when we tried to reingest it :/

**Reply by agrofunk** [reply]
Is there any way of catching the error message to feed a list of bad links and therefore, run the search again with a period ending right before the bad link date? 

Otherwise, one would have to manually list these scenes, which is not a good solution since the bad dates change from path/row to path/row. 

If this process of catching error, listing bad links and running the search again is possible, one can retrieve a whole time-series by concatenating the xarrays skipping the bad links. 

That is my goal!

p.s.: I've been working in many parts of Brazil and for what I noticed was that the missing scenes occur in 2022.

**Reply by TomAugspurger** [reply]
> Is there any way of catching the error message to feed a list of bad links and therefore, run the search again with a period ending right before the bad link date?

It depends a bit on how exactly you're loading the data. Wit something like stackstac or odc.stac, I don't think it would be too easy. The STAC metadata (which is all that those tools use, until you actual read the data) doesn't know that the assets are invalid.

*A* way to check this is to manually iterate over the assets you're using, do a HEAD request on each one, and check the size. If it's abnormally small (like the error message) then you can assume that asset is bad and skip the item. Not a great workaround though :/

**Comment by TomAugspurger** [comment]
We've updated the STAC items for all of the scenes reported in this discussion. Note that for some of the scenes, the actual URLs in blob storage have changed because USGS reprocessed that scene since we originally pulled it and USGS embeds the processing date in the file paths.

For example, the STAC item https://planetarycomputer.microsoft.com/api/stac/v1/collections/landsat-c2-l2/items/LC09_L2SP_174072_20220705_02_T1 previously linked to assets under the prefix

```
https://landsateuwest.blob.core.windows.net/landsat-c2/level-2/standard/oli-tirs/2022/174/072/LC09_L2SP_174072_20220705_20220803_02_T1
```

Now, it links to assets under the prefix

```
https://landsateuwest.blob.core.windows.net/landsat-c2/level-2/standard/oli-tirs/2022/174/072/LC09_L2SP_174072_20220705_20230408_02_T1
```


(note the `20230408`, which indicates the date that scene was reprocessed)

We'll eventually delete the original assets, with the problematic files.

I don't have an ETA yet, but we'll be auditing our landsat archive to find all the scenes with at least one buggy asset, and will re-download each of those scenes.

**Reply by mikemahoney218** [reply]
Thanks Tom, this is fantastic!

**Reply by cullen-molitor** [reply]
Great news! Is this sufficient to mark this as answered or would it be better to keep the thread unanswered until the audit is complete?

**Reply by TomAugspurger** [reply]
Up to you :) Either way, I'll post here once everything is fixed for good.

**Reply by rob-webster-space-intelligence** [reply]
I ran into this issue today with the following asset; one more for the list!

https://landsateuwest.blob.core.windows.net/landsat-c2/level-2/standard/oli-tirs/2022/171/063/LC08_L2SP_171063_20221012_20221020_02_T1/LC08_L2SP_171063_20221012_20221020_02_T1_SR_B7.TIF

**Comment by jessjaco** [comment]
Another one: https://landsateuwest.blob.core.windows.net/landsat-c2/level-2/standard/etm/2022/097/065/LE07_L2SP_097065_20221029_20221124_02_T1/LE07_L2SP_097065_20221029_20221124_02_T1_SR_B5.TIF

**Comment by KBodolai** [comment]
Hello! I've been running into this issue lately. So far I've been listing the offending files and ignoring them. Is it useful for me to post them here, or should I just stay put?

Did you get around to auditing the landsat archive? Anything we can do to help?

**Reply by TomAugspurger** [reply]
Still on our TODO list, sorry. If you want to post a couple Item IDs / bad files feel free to post them and we could explicitly double check those when we do the audit / fix. But no deal if it's a hassle.

**Reply by KBodolai** [reply]
No worries at all, I can imagine how gargantuan a task this is with such a big catalogue!

Below is a list with a few. To filter them out I ended up using the `item.id`, so unfortunately these aren't pointing to the specific asset that's problematic, but probably good for a sense-check. I'll make sure to note down the full link from now on, and add them to the thread :)

<details>

```python
 ['LC08_L2SP_126051_20221102_02_T1',
 'LC08_L2SP_200037_20221108_02_T1',
 'LC08_L2SP_171063_20221012_02_T1', 
 'LC09_L2SP_197055_20220401_02_T1',
 'LC08_L2SP_198055_20221025_02_T1',
 'LC09_L2SP_171063_20220630_02_T1',
 'LC09_L2SP_171064_20220630_02_T1', 
 'LC08_L2SP_172075_20220512_02_T1',
 'LC09_L2SP_172075_20220402_02_T1',
 'LC08_L2SP_172063_20220731_02_T1',
 'LC08_L2SP_172064_20220629_02_T1',
 'LC08_L2SP_172062_20220731_02_T1',
 'LC08_L2SP_186054_20221106_02_T1',
 'LC08_L2SP_189052_20130611_02_T1'] 
```
</details>
